src/mlpack/core/data/split_data.hpp

```diff
@@ -14,9 +14,139 @@
+ * StratifiedSplit(input, label, trainData,
+ *                 testData, trainLabel, testLabel, 0.3);
+ * @endcode
+ *
+ * @param input Input dataset to stratify.
+ * @param inputLabel Input labels to stratify.
+ * @param trainData Matrix to store training data into.
+ * @param testData Matrix to store test data into.
+ * @param trainLabel Vector to store training labels into.
+ * @param testLabel Vector to store test labels into.
+ * @param testRatio Percentage of dataset to use for test set (between 0 and 1).
+ * @param shuffleData If true, the sample order is shuffled; otherwise, each
+ *     sample is visited in linear order. (Default true.)
+ */
+template<typename T, typename U>
+void StratifiedSplit(const arma::Mat<T>& input,
+                     const arma::Row<U>& inputLabel,
+                     arma::Mat<T>& trainData,
+                     arma::Mat<T>& testData,
+                     arma::Row<U>& trainLabel,
+                     arma::Row<U>& testLabel,
+                     const double testRatio,
+                     const bool shuffleData = true)
+{
+  /**
+   * Basic idea:
+   * Let us say we have to stratify a dataset based on labels:
+   * 0 0 0 0 0 1 1 1 1 1 1 (5 0s, 6 1s)
+   *
+   * Let our test ratio be 0.2:
+   * We visit each label and keep the count of each label in our unordered map
+   *
+   * Whenever we encounter a label, we calculate the floored values of
+   * current_count * test_ratio --- labelMap[label] * testRatio and
+   * current_count+1 * test_ratio --- (labelMap[label] + 1) * testRatio
+   *
+   * If there is a difference in counts, then we add to our test set
+   *
+   * Considering our example
+   * 0 -- train set ( 0 * 0.2 == 1 * 0.2 ) (After floor)
+   * 0 -- train set ( 1 * 0.2 == 2 * 0.2 ) (After floor)
+   * 0 -- train set ( 2 * 0.2 == 3 * 0.2 ) (After floor)
+   * 0 -- train set ( 3 * 0.2 == 4 * 0.2 ) (After floor)
+   * 0 -- test set  ( 4 * 0.2 <  5 * 0.2 ) (After floor)
+   *
+   * 1 -- train set ( 0 * 0.2 == 1 * 0.2 ) (After floor)
+   * 1 -- train set ( 1 * 0.2 == 2 * 0.2 ) (After floor)
+   * 1 -- train set ( 2 * 0.2 == 3 * 0.2 ) (After floor)
+   * 1 -- train set ( 3 * 0.2 == 4 * 0.2 ) (After floor)
+   * 1 -- test set  ( 4 * 0.2 <  5 * 0.2 ) (After floor)
+   * 1 -- train set ( 5 * 0.2 == 6 * 0.2 ) (After floor)
+   *
+   * Finally
+   * train set,
+   * 0 0 0 0 1 1 1 1 1 (4 0s, 5 1s)
+   *
+   * test set,
+   * 0 1
+   */
+  arma::uvec indices;
+  indices.set_size(inputLabel.n_cols);
+
+  size_t trainIdx = 0;
+  size_t testIdx = inputLabel.n_cols - 1;
+  std::unordered_map<U, size_t> labelMap;
+
+  arma::uvec order =
+      arma::linspace<arma::uvec>(0, input.n_cols - 1, input.n_cols);
+
+  if (shuffleData)
+  {
+    order = arma::shuffle(order);
+  }
+
+  for (auto i: order)
+  {
+    auto label = inputLabel[i];
+
+    if (floor(labelMap[label] * testRatio) ==
+        floor((labelMap[label] + 1) * testRatio))
+    {
+      indices[trainIdx] = i;
```

Comment:
I think it would be possible to avoid the use of `indices`, if you wanted.  I believe that you can, in the first pass, simply compute the number of training points for each class.  (So instead of `indices`, you'd hold something like `arma::uvec trainingClassCounts`, where at the end of the loop `trainingClassCounts[i]` is the number of training points that are needed from class `i`.)

Then, in the second pass when you extract the data from `input` and `inputLabel` into `trainData` and `trainLabel`, you would also maintain something like `arma::uvec currentClassCounts`, and each time you encounter a point with class `i` where `currentClassCounts[i] <= trainingClassCounts[i]`, you copy the label and the point to the training set and increment `currentClassCounts[i]`.  Otherwise, you copy the label and point to the test set.

Let me know what you think---it's just an idea. :)  It could save a little bit of memory.